app/src/main/java/com/willowtree/vocable/keyboard/KeyboardFragment.kt

```diff
@@ -75,21 +77,40 @@ class KeyboardFragment : BaseFragment() {
 
     private fun populateKeys() {
         KEYS.withIndex().forEach {
-            layoutInflater.inflate(R.layout.keyboard_key_layout, binding?.keyboardKeyHolder, true)
-            (binding?.keyboardKeyHolder?.getChildAt(it.index) as VocableButton).text = it.value
+            with(
+                KeyboardKeyLayoutBinding.inflate(
+                    layoutInflater,
+                    binding?.keyboardKeyHolder,
+                    true
+                ).root as ActionButton
+            ) {
+                text = it.value
+                action = {
+                    val currentText = binding?.keyboardInput?.text?.toString() ?: ""
+                    if (isDefaultTextVisible()) {
+                        binding?.keyboardInput?.text = null
+                        binding?.keyboardInput?.append(text?.toString())
+                    } else if (currentText.endsWith(". ") || currentText.endsWith("? ")) {
+                        binding?.keyboardInput?.append(text?.toString())
+                    } else {
+                        binding?.keyboardInput?.append(text?.toString()?.toLowerCase(Locale.getDefault()))
+                    }
+                }
+            }
         }
     }
 
+    private fun isDefaultTextVisible(): Boolean {
+        if (binding?.keyboardInput?.text.toString() == getString(R.string.keyboard_select_letters)) {
```

Comment:
We could simplify this to just:
`return binding?.keyboardInput?.text.toString() == getString(R.string.keyboard_select_letters)`